revoke/revoke.go

```diff
@@ -123,69 +213,141 @@ func getIssuer(cert *x509.Certificate) *x509.Certificate {
 	}
 
 	return issuer
-
 }
 
-// check a cert against a specific CRL. Returns the same bool pair
-// as revCheck.
-func certIsRevokedCRL(cert *x509.Certificate, url string) (revoked, ok bool) {
-	crl, ok := CRLSet[url]
+// checks whether CRL in memory is valid
+func (r *Revoke) isInMemoryCRLValid(key string) bool {
+	defer r.lock.Unlock()
+	r.lock.Lock()
+	crl, ok := r.crlSet[key]
 	if ok && crl == nil {
 		ok = false
-		delete(CRLSet, url)
+		delete(r.crlSet, key)
 	}
 
-	var shouldFetchCRL = true
-	if ok {
-		if !crl.HasExpired(time.Now()) {
-			shouldFetchCRL = false
+	if ok && !crl.HasExpired(time.Now()) {
+		return true
+	}
+
+	return false
+}
+
+// fetchLocalCRL reads CRL from the local filesystem
+// force flag allows you to update the CRL
+func (r *Revoke) fetchLocalCRL(newLocalCRL string, force bool) error {
+	if force || !r.isInMemoryCRLValid(newLocalCRL) {
+		if _, err := os.Stat(newLocalCRL); err != nil {
+			return fmt.Errorf("failed to read local CRL path: %v", err)
+		}
+
+		tmp, err := ioutil.ReadFile(newLocalCRL)
+		if err != nil {
+			return fmt.Errorf("failed to read local CRL path: %v", err)
+		}
+		crl, err := x509.ParseCRL(tmp)
+		if err != nil {
+			return fmt.Errorf("failed to parse local CRL file: %v", err)
 		}
+
+		r.lock.Lock()
+		r.crlSet[newLocalCRL] = crl
+		if r.localCRL != newLocalCRL {
```

Comment:
User sets invalid path, that's user's fault, user makes this checker into a bad state. I don't understand why you want to revert the change. Trying to be clever/robust against human error is dumb, it makes your code logic hard to be reasoned about. Please, just don't do it, because I won't approve it.